src/main/java/org/codehaus/groovy/classgen/asm/OperandStack.java

```diff
@@ -338,6 +338,12 @@ private void doConvertAndCast(ClassNode targetType, final boolean coerce) {
             return;
         }
 
+        // GROOVY-10034: no need to cast non-primitive array when target is "java.lang.Object[]"
```

Comment:
How about making the check more general? If the `top` component type is inherited from `target` component type, no need to cast type. The following code could be tweaked as:
```java
        if (targetType.isArray() && !ClassHelper.isPrimitiveType(targetType.getComponentType()) 
                  && top.isArray() && !ClassHelper.isPrimitiveType(top.getComponentType())
                  && targetType.getComponentType().isAssignableFrom(top.getComponentType())) {
            return;
        }
```

Reply:
It seems that we could make the check simpler, we do not add extra checks but tweak the existing check:

https://github.com/apache/groovy/blob/master/src/main/java/org/codehaus/groovy/classgen/asm/OperandStack.java#L334
```java
if (targetType == top) return;
```
tweaked as:
```java
if (targetType.getTypeClass().isAssignableFrom(top.getTypeClass())) return;
```

Reply:
I wanted to keep the scope limited.  This is so widely used.  You could certainly add a new issue that references this one and I could take a look at it after I get through with array stuff.

Reply:
`getTypeClass()` is a known unsafe method to use.  I would caution against that without thorough testing.

Reply:
OK. 
Here is the JIRA ticket to track the discusstion: https://issues.apache.org/jira/browse/GROOVY-10035